src/swarm-lang/Swarm/Language/LSP/Hover.hs

```diff
@@ -88,59 +89,79 @@ posToRange myRope foundSloc = do
   -- | parent term
   Syntax' ty ->
   -- | absolute offset within the file
   Int ->
-  Syntax' ty
-narrowToPosition s0@(Syntax' _ t _ ty) pos = fromMaybe s0 $ case t of
-  SLam lv _ s -> d (locVarToSyntax' lv $ getInnerType ty) <|> d s
-  SApp s1 s2 -> d s1 <|> d s2
-  SLet _ _ lv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> d (locVarToSyntax' lv lty) <|> d s1 <|> d s2
-  SBind mlv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> (mlv >>= d . flip locVarToSyntax' (getInnerType lty)) <|> d s1 <|> d s2
-  STydef typ typBody _ti s1 -> d s1 <|> Just (locVarToSyntax' (tdVarName <$> typ) $ fromPoly typBody)
-  SPair s1 s2 -> d s1 <|> d s2
-  SDelay s -> d s
-  SRcd m -> asum . map d . catMaybes . M.elems $ m
-  SProj s1 _ -> d s1
-  SAnnotate s _ -> d s
-  SRequirements _ s -> d s
-  SParens s -> d s
-  -- atoms - return their position and end recursion
-  TUnit -> Nothing
-  TConst {} -> Nothing
-  TDir {} -> Nothing
-  TInt {} -> Nothing
-  TText {} -> Nothing
-  TBool {} -> Nothing
-  TVar {} -> Nothing
-  TStock {} -> Nothing
-  TRequire {} -> Nothing
-  TType {} -> Nothing
-  -- these should not show up in surface language
-  TRef {} -> Nothing
-  TRobot {} -> Nothing
-  TAntiInt {} -> Nothing
-  TAntiText {} -> Nothing
-  TAntiSyn {} -> Nothing
-  SSuspend {} -> Nothing
+  NonEmpty (Syntax' ty)
+pathToPosition s0 pos =
+  case innerPath s0 of
+    Nothing -> s0 :| []
+    Just ss -> s0 :| ss
  where
+  innerPath :: Syntax' ty -> Maybe [Syntax' ty]
+  innerPath (Syntax' _ t _ ty) = case t of
+    SLam lv _ s -> d (locVarToSyntax' lv $ getInnerType ty) <|> d s
+    SApp s1 s2 -> d s1 <|> d s2
+    SLet _ _ lv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> d (locVarToSyntax' lv lty) <|> d s1 <|> d s2
+    SBind mlv _ _ _ s1@(Syntax' _ _ _ lty) s2 -> (mlv >>= d . flip locVarToSyntax' (getInnerType lty)) <|> d s1 <|> d s2
+    STydef typ typBody _ti s1 -> d s1 <|> Just [locVarToSyntax' (tdVarName <$> typ) $ fromPoly typBody]
+    SPair s1 s2 -> d s1 <|> d s2
+    SDelay s -> d s
+    SRcd m -> asum . map d . catMaybes . M.elems $ m
+    SProj s1 _ -> d s1
+    SAnnotate s _ -> d s
+    SRequirements _ s -> d s
+    SParens s -> d s
+    -- atoms - return their position and end recursion
+    TUnit -> mempty
+    TConst {} -> mempty
+    TDir {} -> mempty
+    TInt {} -> mempty
+    TText {} -> mempty
+    TBool {} -> mempty
+    TVar {} -> mempty
+    TStock {} -> mempty
+    TRequire {} -> mempty
+    TType {} -> mempty
+    -- these should not show up in surface language
+    TRef {} -> mempty
+    TRobot {} -> mempty
+    TAntiInt {} -> mempty
+    TAntiText {} -> mempty
+    TAntiSyn {} -> mempty
+    SSuspend {} -> mempty
+
   d = descend pos
+  -- try and decend into the syntax element if it is contained with position
+  descend ::
+    ExplainableType ty =>
+    -- \| position
+    Int ->
+    -- \| next element to inspect
+    Syntax' ty ->
+    Maybe [Syntax' ty]
+  descend p s1@(Syntax' l1 _ _ _) = do
+    guard $ withinBound p l1
+    liftA2 (:) (Just s1) (innerPath s1)
```

Comment:
I prefer `(s1 :) <$> innerPath s1` although it's a matter of taste.  Pick whichever you think is more readable.